### PR TITLE
SPLICE-2285 Basic predicate simplification.

### PR DESCRIPTION
**Pull request by msirek**
Improve the performance of application-generated queries which create dummy predicates such as 1=1, or 4=7, by pruning them out of the WHERE clause or HAVING clause.

For more details, see [SPLICE-2285](https://splice.atlassian.net/browse/SPLICE-2285)

### CODE REVIEW

File: db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/SelectNode.java [hunk 1]
```diff
@@ -526,6 +526,32 @@ public void bindExpressions(FromList fromListParam) throws StandardException{
...
+        // Evaluate expressions with constant operands here to simplify the
+        // query tree and to reduce the runtime cost. Do it before optimize()
+        // since the simpler tree may have more accurate information for
+        // the optimizer. (Example: The selectivity for 1=1 is estimated to
+        // 0.1, whereas the actual selectivity is 1.0. In this step, 1=1 will
+        // be rewritten to TRUE, which is known by the optimizer to have
+        // selectivity 1.0.)
+        // This is also done before predicate simplification to enable
+        // more predicates to be pruned away.
+        accept(new ConstantExpressionVisitor());
```
Comment (yxia92) [inline db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/SelectNode.java]: Should we skip the nested SelectNode here? Given that SelectNode.bindExpressions() is recursively called, it seems that we could visit a nested SelectNode with the ConstantExpressionVisitor multiple times.

File: db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/SelectNode.java [hunk 2]
```diff
@@ -526,6 +526,32 @@ public void bindExpressions(FromList fromListParam) throws StandardException{
...
+        // FALSE OR col1 = 1         ==>  col1 = 1
+        //
+        // Predicate simplification is done before binding because
+        // Subqueries and aggregates in the WHERE clause and HAVING
+        // clause get added to the whereSubquerys, whereAggregates,
+        // havingSubquerys and havingAggregates lists during binding.
+        // If the predicates containing thoses SubqueryNodes and
+        // AggregateNodes are subsequently removed from the WHERE or
+        // HAVING clause, query compilation will fail.
+        if (!getCompilerContext().getDisablePredicateSimplification())
```
Comment (yxia92) [inline db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/SelectNode.java]: Should we restrict it to just WHERE and HAVING clause of a SELECT node?

File: db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/SelectNode.java (hunk 1 above)
Comment (msirek) [reply db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/SelectNode.java]: @yxia92 I agree.  Made the change.

File: db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/SelectNode.java (hunk 2 above)
Comment (msirek) [reply db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/SelectNode.java]: Yes, made the change.

**Comment by msirek** [comment]
@MurrayBrown Default should be on.

**Comment by msirek** [comment]
jenkins please test branch @cdh5.12.2

**Comment by spliceci** [comment]
TEST FAILED

**Comment by msirek** [comment]
jenkins please test branch @cdh5.12.2

**Comment by spliceci** [comment]
TEST FAILED

**Comment by msirek** [comment]
jenkins please test branch @cdh5.12.2

**Comment by spliceci** [comment]
TEST SUCCEEDED +1

**Pull request merged**